Sources/Apodini/Request/Dynamics.swift

```diff
@@ -0,0 +1,100 @@
+//
+//  Dynamics.swift
+//  
+//
+//  Created by Max Obermeier on 10.12.20.
+//
+
+import Foundation
+
+/// Dynamics provides same functionality as `DynamicProperty`, but for elements that are only known
+/// at statup and not compile-time. That is, `Dynamics` stores named elements and makes them
+/// discoverable to the Apodini runtime framework.  This can be used to e.g. delay the decision, which
+/// `Parameter`s are exported to startup-time
+@dynamicMemberLookup
+public struct Dynamics<Element> {
```

Comment:
Similar to the comment made by @nerdsupremacist I would suggest to constraint `Dynamics` or `Injectables` to a base protocol that all injectable elements conform to, such as `RequestInjectable` or a more generalized `Injectable` implementation.
I think that `Injectables` as proposed by @hendesi would fit better to the naming of `RequestInjectable` or `Injectable` and therefore would suggest to rename `Dynamics` to `Injectables`. I think this more clearly details the functionality of injecting multiple different elements conforming to `RequestInjectable` or `Injectable`.

Reply:
In my opinion, `Dynamics` as well as `DynamicProperty` code for making things discoverable to the Apodini framework.

I don't want to constrain Apodini to only be able to discover `RequestInjectable`s (or a more generic `Injectable`s). The only thing I could imagine is constraining it to a `public protocol Discoverable { }` with no requirements that is implemented by all of Apodini's property wrappers.

However, a `DynamicProperty` can also contain properties of any type (and there is nothing we can do about that), so why should we constrain `Dynamics`. It doesn't make any difference for us because we will only search for elements of those types we are interested in. It only gets more complicated for the programmer using Apodini, because they have to use `Dynamics<Any>` and must use type-assertions when using the elements.

Reply:
As detailed in my response to a thread above I think that I did not do a good job explaining my intention behind the `Injectable` protocol. It's vision closely aligns with what you have described as `Discoverable` in the massage above. The response in the thread above (https://github.com/Apodini/Apodini/pull/56/files#r546223303) also aligns with your comment to make all property wrappers conform to that protocol.

Consequently I would also suggest to move to a more descriptive name such as `Injectables` to indicate that it is a collection of `Injectable`s that wrap around properties to inject specific information (e.g. `@Environment` from a global storage, `@State` from a state storage across different requests, `@ObservedObjects` from some global environment or state storage across different requests and mainly `@Parameters` that get information injected from requests). Constraining the content of `@Injectables` to this protocol makes it clear that we expect the developer to add any existing types that can act as property wrappers and e.g. `@PathParemter` that should not conform to `Injectable` or a simple `Int` or `String` that would need to be wrapped by an `@Parameter` property wrapper or is part of a dictionary literal expressing it as `Parameter<Int>()`.

My and I suspect also @nerdsupremacist 's intention behind this move was to use to type system to hint to the developer using Apodini what the developer can use as part of the `@Injectables`/`@Dynamics` property wrapper.   Some custom `Injectable`s/`Dynamic`s can of course have properties that are not marked with an Apodini or custom property wrapper but the enclosing type must conform to `Injectable`.


Regarding your last comment about type erasure using Any: I don't think it will be necessary to make `Injectables`/`Dynamics` have a generic type and just use the `Injectable` protocol in the type definitions:
```suggestion
public struct Injectables {
    internal var elements: [String: Injectables]
```
You could have a typed version like `TypedInjectables<Element: Injectable>` that provides you a concretely typed version of an `Injectables` instance when calling the `typed` method defined below. This would also enable you to reduce the need of `Any` in your test cases as far as I can see the implications on the examples in the test cases below.